python_modules/dagster/dagster/core/definitions.py

```diff
@@ -337,27 +432,48 @@ def __init__(
             key_type=str,
             value_type=PipelineContextDefinition,
         )
+        dependencies = check_opt_two_dim_dict(
+            dependencies,
+            'dependencies',
+            value_type=DependencyDefinition,
+        )
+
+        processed_dependencies = {}
+        solid_aliases = defaultdict(list)
+
+        for solid_key, definition in dependencies.items():
+            if isinstance(solid_key, Solid):
+                if solid_key.alias:
+                    solid_aliases[solid_key.name].append(solid_key.alias)
+                    processed_dependencies[solid_key.alias] = definition
+                else:
+                    processed_dependencies[solid_key.name] = definition
+            else:
+                processed_dependencies[solid_key] = definition
 
+        pipeline_solids = []
         for solid in solids:
-            if not isinstance(solid, SolidDefinition) and callable(solid):
+            if isinstance(solid, SolidDefinition):
+                for alias in solid_aliases[solid.name]:
+                    pipeline_solids.append(PipelineSolid(name=alias, definition=solid))
+                pipeline_solids.append(PipelineSolid(name=solid.name, definition=solid))
+            elif callable(solid):
                 raise DagsterInvalidDefinitionError(
                     '''You have passed a lambda or function {func} into
-                a pipeline that is not a solid. You have likely forgetten to annotate this function
-                with an @solid or @lambda_solid decorator located in dagster.core.decorators
-                '''.format(func=solid.__name__)
+                    a pipeline that is not a solid. You have likely forgetten to annotate this function
+                    with an @solid or @lambda_solid decorator located in dagster.core.decorators
+                    '''.format(func=solid.__name__)
                 )
+            else:
+                check.list_param(solids, 'solids', SolidDefinition)
```

Comment:
You are doing this check N times

Reply:
It will fail on the first one that's not a solid definition, no?

Reply:
I know it's not super efficient, but it will only do it in error case anyway.

Reply:
good point.